### ISSUE TITLE
Default for uidMappings / gidMappings if not provided in rootless mode

### ISSUE THREAD

**Issue by olifre** [opencontainers/runc#1714]
Looking at OCI, `uidMappings` and `gidMappings` should be optional, but it seems `runc` requires them when running in rootless mode:
```
$ runc --root /tmp/runc run somefun                                                                                                                                                                        
failed to get root uid from uidMappings: User namespaces enabled, but no uid mappings found.
```
It should be sensible to choose the uid / gid of the invoking user and map `size 1` in case nothing is provided. This would make an OCI container's `config.json` usable by several different users without the need to keep a separate `config.json` per user.

**Comment by cyphar** [comment]
As with the read-only issue we've discussed separately, my main problem with this is two-fold:

* It adds more out-of-spec features which make it harder for people to implement the OCI specification (as they end up having to mirror `runc`'s actual semantics rather than the ones dictated by the spec).

* This would actually violate the specification. The spec actually says that `uidMappings` and `gidMappings` are only allowed if you have user namespaces enabled (and in this mode they are actually required). I'm not sure that we actually would be allowed to have out-of-spec defaults like that.

I do agree in principle that only having a single `config.json` per-user is an issue, but this concern doesn't make too much sense in the context of rootless containers (especially when you've extracted the rootfs as the user in question -- trying to run a rootless container bundle with a different user won't work).

The more general concern of how bundles are designed is a more overarching problem that I've seen several people (including @LittleLightLittleFire) complain about. I'm not sure what exactly would be the best way of dealing with that issue either...

**Comment by olifre** [comment]
@cyphar Understood. So from this, I would deduce that the "issue" I raised is rather an issue with the specs - since they actively prevent sharing `config.json` amongst users. 

I must however disagree that this concern does not make too much sense in the context of rootless containers. The usecase I have mentioned would be running thousands of containers e.g. at HPC clusters or other computing sites, which would also prefer to run containers unprivileged. Often, the cluster's workload management system takes care to distribute the containers - and extracting them each single time for each single job would be a huge waste of resources, so shared filesystems are used, and CVMFS is one of those ideally suited for container delivery. So in short, it's not unlikely that the majority of containers running at a single point in time world-wide would run in such a configuration, which is why I would not say `it does not make much sense`. 

A workaround that would help would be if the workload management system "on the fly" created a `config.json` per job. I don't know of any workload management system doing it yet, but after all, rootless containers are still pretty young. 

Of course, for single users you are perfectly right - then it does not really make sense that `config.json` can not be shared for rootless containers. 

So maybe, this issue should be raised against the specification and the design of bundles, since they make life more complicated for a large number of users, and fixing that would likely be easy. 
Which place is the correct place to complain to (or rather, fire off a discussion) on this topic?

**Comment by cyphar** [comment]
> Often, the cluster's workload management system takes care to distribute the containers - and extracting them each single time for each single job would be a huge waste of resources, so shared filesystems are used, and CVMFS is one of those ideally suited for container delivery.

I agree (and I'm trying to figure out ways of handling the sharing problem as an unprivileged user) -- but does this mean that CVMFS is faking the UID/GID of files in the rootfs in order to allow for this to work? If you're already faking the UID/GID you could also generate new `config.json`s on-the-fly no (of course this is uglier than necessary, just wondering)?

> Which place is the correct place to complain to (or rather, fire off a discussion) on this topic?

https://github.com/opencontainers/runtime-spec -- but you might want to also send an email to dev@opencontainers.org to make sure that folks see it.

**Comment by olifre** [comment]
> [...] but does this mean that CVMFS is faking the UID/GID of files in the rootfs in order to allow for this to work?

No - the files are (usually) visible on the client as the same user who put the files in CVMFS, which in our case is `root`. This is not an issue, though - since the container is read-only anyways, and the "host root user" is not mapped, the files will show up as owned by nobody in the user-namespace / rootless container, and be readable just fine. No faking is needed or performed. 

Many thanks for the links, I'll create an issue with them and send a mail in a calm moment, and for sure reference it here to keep everybody up-to-date!

**Comment by olifre** [comment]
@cyphar You said:

> The spec actually says that uidMappings and gidMappings are only allowed if you have user namespaces enabled (and in this mode they are actually required).

I don't manage to find the correct place in the specs which says this - all I find is
https://github.com/opencontainers/runtime-spec/blob/fa4b36aa9c99e00c2ef7b5c0013c84100ede5f4e/config-linux.md#user-namespace-mappings
which only claims they should be optional. Could you help me in finding the correct place in the runtime-specs? Many thanks!

**Comment by cyphar** [comment]
> Could you help me in finding the correct place in the runtime-specs? Many thanks!

Oh. I think you're right. Now, the million dollar question is was this intentional (so we could implement this) or was it accidentally broken during the refactor in opencontainers/runtime-spec@621684f6450a1e1776febd62a5df32c35a793fce (in which case we should probably fix it or discuss whether we're happy with the current version). It has always been my understanding that the mapping arguments are required.

But in that case, yes we could actually remove the need for `uidMappings` and `gidMappings`. However, the main downside of adding this is that we might need to add a `--rootless` flag (as with #1692) since we will no longer be able to "just try" the configuration (though we could just check if `.linux.namespaces` contains `user` and it doesn't have any mappings added).

/cc @crosbymichael WDYT?

**Comment by crosbymichael** [comment]
From the spec POV, the mappings are optional.  If you add additional settings like, `CLONE_NEWUSER` then mappings become required to fulfill the request, just like many other fields like `UTS` and `hostname`. 

I think the current functionality is correct and we don't want to have runc generate these things at runtime, the spec should be complete and not modified at runtime. 

Spec field requirements/optional is about schema not kernel requirements.

**Comment by olifre** [comment]
I think the remaining question then is: If the schema does not require it, and the kernel requires it - does that necessarily mean that the implementation should enforce it to be present in the schema? Or is it fine that the implementation chooses sensible defaults if a non-mandatory field is not provided? 

My main issue with having to specify those mappings is that this means shared containers / containers on a shared filesystem for use by multiple users (which is a growing usecase!) are rendered impossible, unless a separate tooling to generate separate `config.json` for each user on-the-fly is used. Is this really the way to proceed? 
The mappings seem to be the only thing preventing this usecase. 

An alternative idea (which would however mean a change to the spec) could be to have something like this work:
```
                "uidMappings": [
                        {
                                "hostID": -1,
                                "containerID": 0,
                                "size": 1
                        }
                ],
                "gidMappings": [
                        {
                                "hostID": -1,
                                "containerID": 0,
                                "size": 1
                        }
                ],
```
i.e. if `hostID` is set to `-1` (or whatever is more reasonable), the runtime should use the UID / GID of the user on the host. This would fix the issue of not being able to share containers + `config.json`, and still be technically complete (I think). 

Other opinions?

**Comment by crosbymichael** [comment]
The runtime spec was not designed as a distribution artifact and that is why you are seeing issues like this.  It was designed to be generated by tooling at runtime.

Also the `--rootless` flag is not a runtime attribute, it's only for spec generation in runc, the spec is what makes a container rootless or not.  runc treats the spec literally so I don't think having a `-1` on the mappings will be an acceptable solution.

**Comment by olifre** [comment]
> The runtime spec was not designed as a distribution artifact and that is why you are seeing issues like this. It was designed to be generated by tooling at runtime.

This was not clear to me, indeed! My belief was that the spec was also meant to be a distribution artifact. I must wonder how rootless bundles will work in that case, when `config.json` is part of the distribution (if I understood correctly), but cannot be used without prior change of the UID / GID mappings. 

In this case, indeed my proposal is naught - and all tools which want to leverage OCI containers should create specfiles / re-use existing tooling to do so. This means workload management systems such as HTCondor, Slurm, WLCG Compute Elements and many more should learn to create temporary, per-job specfiles to leverage runc. Most of those only have basic support for Docker at the moment, but I'm not aware of any workload management system as of yet supporting OCI - looking forward to that.
